function timeout now disposes of WebHost services

brettsam opened the pull request:
Fixes https://github.com/Azure/azure-functions-host/issues/5341

There were two distinct problems here:
1. We were never disposing our root container, meaning metrics from the MetricsEventManager were not flushed.
2. On timeout (and elsewhere) we were using the incorrect `IApplicationLifetime` interface -- we need the `Microsoft.AspNetCore.Hosting` one, not the `Microsoft.Extensions.Hosting`. This was recently introduced in #5437 and hasn't gone out yet.

brettsam commented on src/WebJobs.Script.WebHost/DependencyInjection/WebHostServiceProvider.cs:
```diff
@@ -50,5 +50,10 @@ public IServiceScope CreateScope()
+
+        public void Dispose()
+        {
+            _container?.Dispose();
```
This is the core issue...

brettsam commented on src/WebJobs.Script.WebHost/FileMonitoringService.cs:
```diff
@@ -24,7 +24,7 @@ public class FileMonitoringService : IHostedService, IDisposable
         private readonly ScriptJobHostOptions _scriptOptions;
         private readonly IScriptEventManager _eventManager;
-        private readonly IApplicationLifetime _applicationLifetime;
+        private readonly AspNetCore.Hosting.IApplicationLifetime _applicationLifetime;
```
Lots of namespace updates here -- I've checked and we don't use `Microsoft.Extensions.Hosting.IApplicationLifetime` anywhere anymore.

mathewc replied:
It appears that this interface is now obsolete https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.hosting.iapplicationlifetime?view=aspnetcore-3.1. They say we should be using Microsoft.Extensions.Hosting.IHostApplicationLifetime now?

brettsam replied:
That's new for 3.0 -- but for now the old interface is supported as well. They have their `ApplicationLifetime` class implement both the old and the new: https://github.com/dotnet/aspnetcore/blob/master/src/Hosting/Hosting/src/Internal/ApplicationLifetime.cs#L15

To take a dependency on the new interface would require some changes in 3.0. They'd be good and make this less subtle -- we'd have to explicitly make sure we're calling `StopApplication()` on the WebHost so it'd be clearer what we're doing. But I think we should handle that separately...

fabiocav replied:
Added a comment at the PR level. We should stick to `IApplicationLifetime` for now for the reasons outlined there.

brettsam commented on test/WebJobs.Script.Tests.Integration/Host/WebScriptHostManagerTimeoutTests.cs:
```diff
@@ -4,137 +4,116 @@
-using WebJobs.Script.Tests;
 using Xunit;
 
 namespace Microsoft.Azure.WebJobs.Script.Tests.Host
```
These tests have never run in v2. I've updated them so they're running again and allowed me to validate this E2E.

mathewc replied:
Awesome, thanks

brettsam commented on test/WebJobs.Script.Tests.Integration/Host/WebScriptHostManagerTimeoutTests.cs:
```diff
@@ -4,137 +4,116 @@
             }
+
+            // Validates a bug where WebHost services were not being disposed on a function timeout
+            Assert.True(_disposedService.IsDisposed, "Expected services to be disposed");
```
This test failed before fix was applied.

brettsam commented on test/WebJobs.Script.Tests.Integration/TestFunctionHost.cs:
```diff
@@ -112,6 +112,11 @@ public class TestFunctionHost : IDisposable
+
+            // Wire up StopApplication calls as they behave in hosted scenarios
+            var lifetime = WebHostServices.GetService<AspNetCore.Hosting.IApplicationLifetime>();
+            lifetime.ApplicationStopping.Register(async () => await _testServer.Host.StopAsync());
```
Calling `StopApplication()` in the `TestServer` doesn't work automatically. This makes any call to `StopApplication()` translate to a call to stop the host.

brettsam commented:
@fabiocav -- enjoy :-)

brettsam commented on src/WebJobs.Script.WebHost/FileMonitoringService.cs:
```diff
@@ -85,6 +87,7 @@ public Task StartAsync(CancellationToken cancellationToken)
 
         public Task StopAsync(CancellationToken cancellationToken)
         {
+            StopFileWatchers();
```
This change highlighted an interesting scenario -- when we stop the WebHost, it will also stop the running JobHost. However, if that JobHost.Stop() call takes too long (currently 10 seconds), we may not ever actually *dispose* the JobHost. This means that some services may take WebHost dependencies... which get disposed before the JobHost.

Everything is shutting down at that point so it may not matter, but in this scenario I wanted to try to stop these watchers when we're stopping the host -- I think it makes logical sense anyway.

ankitkumarr commented on src/WebJobs.Script.WebHost/FileMonitoringService.cs:
```diff
@@ -120,6 +123,38 @@ private void InitializeFileWatchers()
 
+        private void StopFileWatchers()
+        {
+            lock (_stopWatchersLock)
```
While at it, would you mind also adding a log entry here saying that file watchers are being stopped?

There were a couple of scenarios where FileWatchers were disposed and it was hard to figure out when debugging through logs.

mathewc commented on src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs:
```diff
@@ -120,12 +120,12 @@ public async void WorkerRestart(HttpWorkerRestartEvent workerRestart)
-                    (_httpWorkerChannel as IDisposable)?.Dispose();
-                }
-                await RestartWorkerChannel(workerId);
+            _logger.LogDebug("Disposing channel for workerId: {channelId}", workerId);
```
Why is the format name for workerId channelId?

fabiocav replied:
Guessing a refactoring leftover :/

brettsam replied:
And I didn't actually change that... this was a spacing auto-cleanup that happened on save.

mathewc commented:
Changes look good overall - my main question is whether we shouldn't be using Microsoft.Extensions.Hosting.IHostApplicationLifetime as the doc recommends, due to the other being obsolete

fabiocav commented:
@mathewc changing the code to use `IHostApplicationLifetime` would cause divergence between 2.0 and 3.0 and make that changes more difficult to manage. Since this will not be removed in the .NET Core major versions either of them depend on, I think we should keep the current approach as the ApplicationLifetime implements the appropriate interfaces in both versions.

fabiocav commented on src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs:
```diff
@@ -303,6 +305,8 @@ protected internal virtual void WriteMetricEvents(SystemMetricEvent[] metricEven
 
         protected virtual void Dispose(bool disposing)
         {
+            _logger.LogDebug($"Disposing {nameof(MetricsEventManager)}");
```
If the factory that creates this logger is disposed prior to this type, we might run into issues here, right? Not sure how likely this is to happen, but it might be good to be defensive.

brettsam replied:
The LoggerFactory (and it's LoggerProviders) will be disposed after this as this takes a dependency on it via the constructor. But I can wrap it in try/catch to be extra careful.

fabiocav commented on src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs:
```diff
@@ -24,7 +24,7 @@ namespace Microsoft.Azure.WebJobs.Script.WebHost
     public class WebJobsScriptHostService : IHostedService, IScriptHostManager, IServiceProvider, IDisposable
     {
-        private readonly IApplicationLifetime _applicationLifetime;
+        private readonly AspNetCore.Hosting.IApplicationLifetime _applicationLifetime;
```
Can we just alias this to avoid the namespaced type name throughout?

brettsam replied:
Yes. I did that in a few spots. Not sure why I didn't do it here. I'll update it to be consistent.

brettsam commented:
@fabiocav / @ankitkumarr -- I've addressed your comments but need another approval before I can merge. Let me know if you'd like to see any other changes.
